text/3712-box-yesalias.md

```diff
@@ -0,0 +1,127 @@
+[^3]: Since [rust#122018](https://github.com/rust-lang/rust/pull/122018), this is fixed as only `Box<T>` has the special opsem behaviour described here, which is the second solution.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+[`alloc::boxed::Box<T>`] is defined with a number of magic features that interact directly with the language. To limit the impact, both on users and on the language itself, we remove the language level requirement that `Box` be dereferenceable and unique (like applies to `&mut T`). 
+This does not affect library level requirements of uniqueness however - it may remain undefined behaviour to access a `Box` that is being aliased (particularily where the operations produce multiple aliasing mutable references from different `Box`es) or to use `Box::from_raw` to construct multiple aliasing boxes. 
+
+We also do not remove any other language level magic from `Box`, such as the ability to do both partial and complete moves from a `Box`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+For the remainder of this section, let `WellFormed<T>` designate a type for *exposition purposes only* defined as follows:
+```rust
+#[repr(transparent)]
+struct WellFormed<T: ?Sized>(core::ptr::NonNull<T>);
+```
+
+(Note that we do not define this type in the public standard library interface, though an implementation of the standard library could define the type locally)
+
+The following are not valid values of type `WellFormed<T>`, and a typed copy that would produce such a value is undefined behaviour:
+* Any value that is invalid for a raw pointer to `T` (e.g. a value read from uninitialized memory, or a fat pointer with invalid metadata)
+* A null pointer (even when `T` is a Zero-sized type)
+* A pointer with an address that is not well-aligned for `T` (or in the case of a DST, the `align_of_val_raw` of the value), or
+* A pointer with an address that offsetting that address (as though by `.wrapping_byte_offset`) by `size_of_val_raw` bytes would wrap arround the address space 
+
+The [`alloc::boxed::Box<T>`] type shall be laid out as though a `repr(transparent)` struct containing a field of type `WellFormed<T>`. The behaviour of doing a typed copy as type [`alloc::boxed::Box<T>`] shall be the same as though a typed copy of the struct `#[repr(transparent)] struct Box<T>(WellFormed<T>);`. 
```

Comment:
This looks to be a new guarantee? We currently don't normatively say anything about the layout of `Box`, AFAIK.

Also see [this](https://internals.rust-lang.org/t/layout-of-dst-box/21728/1); Cc @CAD97

Reply:
Stdlib docs promise it has the same layout and abi as a pointer, and also promises niche optimization.

https://doc.rust-lang.org/alloc/boxed/index.html#memory-layout

> So long as T: Sized, a `Box<T>` is guaranteed to be represented as a single pointer and is also ABI-compatible with C pointers (i.e. the C type `T*`).

https://doc.rust-lang.org/core/primitive.fn.html#abi-compatibility

> `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>` (specifically, only `Box<T, Global>`), and `NonNull<T>` are all ABI-compatible with each other for all T. They are also ABI-compatible with each other for different T if they have the same metadata type (`<T as Pointee>::Metadata`).

(The last sentence implies the abi, and thus layout, holds for `T: !Sized` pointers as well)

Reply:
Ah okay :)